### PR TITLE
deprecate AsdfInFits

### LINKED ISSUE #1332: Add deprecation for AsdfInFits
The ability to read AsdfInFits files is being moved from [ASDF](https://github.com/asdf-format/asdf) into [stdatamodels](https://github.com/spacetelescope/stdatamodels) (in part to consolidate non-pipeline code for jwst). 

AsdfInFits support has been merged into the development branch of stdatamodels: https://github.com/spacetelescope/stdatamodels/pull/110

A deprecation warning or warnings need to be added to ASDF prior to the 2.15 release as this is the last planned release that will support AsdfInFits (we will be moving development to ASDF 3.0 in which we will be dropping AsdfInFits).

### PR DESCRIPTION
**Pull request by braingram**
Support to reading and writing ASDF in fits is moving to [stdatamodels](https://github.com/spacetelescope/stdatamodels).

https://github.com/spacetelescope/stdatamodels/pull/110

fixes #1332

### CODE REVIEW

**Comment by braingram** [comment]
@WilliamJamieson could I get your advice on handling including this deprecation?

specutils is failing because it turns warnings into errors:
https://github.com/astropy/specutils/blob/34b5e84fd107f31a3c2fe2d1aaf95517d68d95f1/setup.cfg#L61

How would you recommend we handling the error issued by this new deprecation warning?

I could open a PR with specutils ignoring the new warning (adding the following to setup.cfg):
```
ignore:ASDFInFits has been deprecated .*:asdf.exceptions.AsdfDeprecationWarning:asdf.fits_embed
```
as long as we're happy with the proposed changes in this PR. We could then wait until that PR is merged to merge this PR.

Is there an easier/better way to handle this? Thanks!

**Comment by WilliamJamieson** [comment]
> @WilliamJamieson could I get your advice on handling including this deprecation?
> 

As far as depreciation warnings are concerned, it is not up to us to decide how `specutils` should handle them. If we want to be nice (and we do), we can open an issue there warning them that we are planning on deprecating something they are using with a timeline for when the deprecation will appear (and when the feature will be removed) alongside with recommendations on how to remediate the issue.

As far as far as actually merging this PR we should wait until `stdatamodels` has a release which contains the changes in spacetelescope/stdatamodels#110. This is because the remediation recommendation is use ASDF-in-fits from `stdatamodels`.

> specutils is failing because it turns warnings into errors: [astropy/specutils@`34b5e84`/setup.cfg#L61](https://github.com/astropy/specutils/blob/34b5e84fd107f31a3c2fe2d1aaf95517d68d95f1/setup.cfg#L61)
> 
> How would you recommend we handling the error issued by this new deprecation warning?
> 
> I could open a PR with specutils ignoring the new warning (adding the following to setup.cfg):
> 
> ```
> ignore:ASDFInFits has been deprecated .*:asdf.exceptions.AsdfDeprecationWarning:asdf.fits_embed
> ```

This is not a good idea. I would rather see a warning capture/ignore for this on a per-test basis as they will likely need to stride testing multiple version requirements.

File: asdf/fits_embed.py
Comment (WilliamJamieson) [inline asdf/fits_embed.py:156]: This whole module is being deprecated, not just his class. As such the warning should be at the module level.

**Review by WilliamJamieson — CHANGES_REQUESTED** [review]

File: asdf/fits_embed.py
Comment (braingram) [reply asdf/fits_embed.py]: Thanks. I moved the warning and added notes and deprecations for asdftool extract and remove_hdu

**Comment by WilliamJamieson** [comment]
@MeeseeksDev Hello

**Comment by braingram** [comment]
There are a few things that I'm not pleased with but require updating other packages to fix.

1) The asdf_in_fits document in stdatamodels does not appear to have a correctly formatted sphinx link (it is missing the trailing colon so it does not appear as a std:label in the object inventory):
https://raw.githubusercontent.com/spacetelescope/stdatamodels/master/docs/source/asdf_in_fits.rst
This made linking to the docs difficult. I opened a PR to fix this (https://github.com/spacetelescope/stdatamodels/pull/122) but don't think it warrants a new release.
2) The custom.css in sphinx_asdf resizes warnings when they become active:
https://github.com/asdf-format/sphinx-asdf/blob/master/sphinx_asdf/static/custom.css#L7
This makes any contained linked unclickable. So for now there is a `note` instead of a `warning` about the asdf in fits deprecation

**Comment by braingram** [comment]
I opened a separate issue to track the deprecation of asdftool commands that require asdf.fits_embed. https://github.com/asdf-format/asdf/issues/1382

**Comment by WilliamJamieson** [comment]
> There are a few things that I'm not pleased with but require updating other packages to fix.
> 
> 1. The asdf_in_fits document in stdatamodels does not appear to have a correctly formatted sphinx link (it is missing the trailing colon so it does not appear as a std:label in the object inventory):
>    [spacetelescope/stdatamodels@`master`/docs/source/asdf_in_fits.rst (raw)](https://raw.githubusercontent.com/spacetelescope/stdatamodels/master/docs/source/asdf_in_fits.rst)
>    This made linking to the docs difficult. I opened a PR to fix this ([fix sphinx link for asdf in fits documentation spacetelescope/stdatamodels#122](https://github.com/spacetelescope/stdatamodels/pull/122)) but don't think it warrants a new release.
> 2. The custom.css in sphinx_asdf resizes warnings when they become active:
>    [asdf-format/sphinx-asdf@`master`/sphinx_asdf/static/custom.css#L7](https://github.com/asdf-format/sphinx-asdf/blob/master/sphinx_asdf/static/custom.css?rgh-link-date=2023-02-16T21%3A23%3A11Z#L7)
>    This makes any contained linked unclickable. So for now there is a `note` instead of a `warning` about the asdf in fits deprecation

In light of these issues, I think we should wait until `sphinx_asdf` is released with fixes for these issues.

**Comment by braingram** [comment]
> > 2. The custom.css in sphinx_asdf resizes warnings when they become active:
> >    [asdf-format/sphinx-asdf@`master`/sphinx_asdf/static/custom.css#L7](https://github.com/asdf-format/sphinx-asdf/blob/master/sphinx_asdf/static/custom.css?rgh-link-date=2023-02-16T21%3A23%3A11Z#L7)
> >    This makes any contained linked unclickable. So for now there is a `note` instead of a `warning` about the asdf in fits deprecation
> 
> In light of this issue, I think we should wait until `sphinx_asdf` is released with fixes for these issues.

I am ok with this as a note and would prefer to have it reviewed.

File: docs/asdf/features.rst
Comment (WilliamJamieson) [inline docs/asdf/features.rst:422]: This should be an intersphinx link. Is this the issue the intersphinx link that will eventually be resolved by spacetelescope/stdatamodels#122? If so we should create an issue to track updating this to a proper link when `stdatamodels` gets fixed.

File: docs/asdf/features.rst
Comment (WilliamJamieson) [inline docs/asdf/features.rst:423]:
```suggestion
.. warning::

    `~asdf.fits_embed.AsdfInFits` and `~asdf.fits_embed` are deprecated and
    will be removed from an upcoming version of ASDF. Support for reading
    and writing ASDF data within a FITS file has moved to the `stdatamodels`
    package. Please see the the `stdatamodels`
    :external+stdatamodels:doc:`asdf_in_fits` documentation for information
    about migrating code from ASDF to `stdatamodels`.
```

Per https://github.com/asdf-format/asdf/pull/1337#issuecomment-1433731516 this should be a warning not a note.

**Review by WilliamJamieson — CHANGES_REQUESTED** [review]
I called out some specific points for #1382, and made some suggestions and other comments.

File: asdf/commands/extract.py
Comment (braingram) [reply asdf/commands/extract.py:48]: Yes, I figured a separate PR would be called for. If you'd rather see those changes included here let me know.

File: asdf/commands/remove_hdu.py
Comment (braingram) [reply asdf/commands/remove_hdu.py:44]: Yup, same as the other comment.

File: asdf/commands/tests/test_extract.py
Comment (braingram) [reply asdf/commands/tests/test_extract.py:20]: I added a comment to the code describing the need for this import.

In short, calling `del sys.modules['asdf.fits_embed']` isn't sufficient to allow for re-triggering the deprecation warning if an import like `from asdf.fits_embed import AsdfInFits` is used).

I chose adding the `import asdf.fits_embed` option to avoid changing the usage of AsdfInFits in test_extract.py but that should work as well (switching it to `asdf.fits_embed.AsdfInFits`). If you're prefer that alternative let me know.

File: asdf/commands/tests/test_extract.py
Comment (WilliamJamieson) [inline asdf/commands/tests/test_extract.py:18]: Given this comment, it seems like this module must be run first in order for the other tests to properly function? 

Tests should be independent of order of execution. If this needs to be done in this module then it must be done in every module expecting a similar warning so that they can be run independently.

File: docs/asdf/features.rst
Comment (braingram) [reply docs/asdf/features.rst:422]: Thanks. I opened an issue here:
https://github.com/asdf-format/asdf/issues/1385

**Review by WilliamJamieson — CHANGES_REQUESTED** [review]

File: asdf/commands/tests/test_extract.py
Comment (WilliamJamieson) [reply asdf/commands/tests/test_extract.py:20]: See https://github.com/asdf-format/asdf/pull/1337#discussion_r1110042655.

File: asdf/commands/tests/test_extract.py
Comment (braingram) [reply asdf/commands/tests/test_extract.py:18]: All of the fits_embed imports in the tests have a similar check to allow for independent order of execution.

File: docs/asdf/features.rst
Comment (braingram) [reply docs/asdf/features.rst:423]: > I am ok with this as a note and would prefer to have it reviewed.

Changing it to a warning makes the links unclickable. I would prefer to leave this as a note.

Comment (WilliamJamieson) [reply docs/asdf/features.rst:423]: We have gone to all the trouble to update `sphinx_asdf` to fix this issue. If you are fine with it being a note then why did we bother with all that? Especially since `sphinx_asdf` is something which is relatively easy to release a patch for.

File: asdf/commands/tests/test_extract.py
Comment (WilliamJamieson) [reply asdf/commands/tests/test_extract.py:18]: I'm not taking about the check I'm talking about the `import asdf.fts_embed`.

File: asdf/commands/tests/test_remove_hdu.py
Comment (WilliamJamieson) [inline asdf/commands/tests/test_remove_hdu.py:14]: Given all the discussion about `import asdf.fits_embed` shouldn't this be:
```suggestion
with pytest.warns(AsdfDeprecationWarning, match="AsdfInFits has been deprecated.*"):
    if "asdf.fits_embed" in sys.modules:
        del sys.modules["asdf.fits_embed"]
    import asdf.fitest_embed
    from asdf.fits_embed import AsdfInFits
```

**Review by WilliamJamieson — CHANGES_REQUESTED** [review]

File: docs/asdf/features.rst
Comment (braingram) [reply docs/asdf/features.rst:423]: I will change it to a warning.

**Comment by WilliamJamieson** [comment]
> I am ok with this as a note and would prefer to have it reviewed.

I wanted to mark it as a draft to avoid an accidental merger until `sphinx_asdf` is released with a fix. I can still review things when it is a draft, it just won't be fully ready until the `sphinx_asdf` release.

File: asdf/commands/tests/test_extract.py
Comment (braingram) [reply asdf/commands/tests/test_extract.py]: I changed the test to use `asdf.fits_embed` instead of `from asdf.fits_embed import AsdfInFits` and removed the comment (as the issues with removing the module caused by the `from` import are no longer relevant).

File: asdf/commands/tests/test_remove_hdu.py
Comment (braingram) [reply asdf/commands/tests/test_remove_hdu.py]: Similar to https://github.com/asdf-format/asdf/pull/1337#discussion_r1110074571 I changed the imports in this test to avoid the confusion.

**Comment by braingram** [comment]
> I wanted to mark it as a draft to avoid an accidental merger until `sphinx_asdf` is released with a fix. I can still review things when it is a draft, it just won't be fully ready until the `sphinx_asdf` release.

@WilliamJamieson I believe I've addressed all of your comments except for this one. Am I correct in thinking that the inclusion of this warning (even as a note) means this PR has to wait for a release of `sphinx_asdf` before merging?

Would it be an acceptable alternative to include the deprecation description as a note and open an issue to convert it to a warning when `sphinx_asdf` is released?

**Comment by WilliamJamieson** [comment]
> @WilliamJamieson I believe I've addressed all of your comments except for this one. Am I correct in thinking that the inclusion of this warning (even as a note) means this PR has to wait for a release of `sphinx_asdf` before merging?

Yes. Although, the only thing holding up a `sphinx_asdf` release are your reviews of

- asdf-format/sphinx-asdf#37
- asdf-format/sphinx-asdf#54

> 
> Would it be an acceptable alternative to include the deprecation description as a note and open an issue to convert it to a warning when `sphinx_asdf` is released?

This would be fine.

**Comment by braingram** [comment]
> > @WilliamJamieson I believe I've addressed all of your comments except for this one. Am I correct in thinking that the inclusion of this warning (even as a note) means this PR has to wait for a release of `sphinx_asdf` before merging?
> 
> Yes. Although, the only thing holding up a `sphinx_asdf` release are your reviews of
> 
>     * [Move to pyproject.toml file sphinx-asdf#37](https://github.com/asdf-format/sphinx-asdf/pull/37)
> 
>     * [Move to using private block API sphinx-asdf#54](https://github.com/asdf-format/sphinx-asdf/pull/54)
> 

Those look unrelated to the removal of custom.css. Are you waiting for both of those before making a release? I doubt I will have a chance to review either before next week at the earliest.

> > Would it be an acceptable alternative to include the deprecation description as a note and open an issue to convert it to a warning when `sphinx_asdf` is released?
> 
> This would be fine.

Sounds great. I opened an issue to track changing the note back to a warning: https://github.com/asdf-format/asdf/issues/1386

**Comment by WilliamJamieson** [comment]
@meekseeksdev backport 2.15.x

**Comment by WilliamJamieson** [comment]
@meeseeksdev backport to 2.15.x

**Pull request merged**
